### PR TITLE
created classes: Stack,StackStatistics

### PR DESCRIPTION
**Pull request by TatsianaGouteeva**
Hi!
Please, check my class Stack and StackStatistics.

### CODE REVIEW

File: Cocoaheads School_Lesson2.playground/Contents.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        return topElement
+    }
+    
+    func top() -> Int? {
+        return array.last
+    }
+}
+
+class StackStatistics: Stack {
+
```
Comment (dersim-davaod) [inline Cocoaheads School_Lesson2.playground/Contents.swift]: nit:
```suggestion
```

File: Cocoaheads School_Lesson2.playground/Contents.swift [hunk 2]
```diff
@@ -0,0 +1,31 @@
+class Stack {
+    var array = [Int]()
+    var minElement: Int?
+    
+    func push(element: Int) {
```
Comment (dersim-davaod) [inline Cocoaheads School_Lesson2.playground/Contents.swift]: Your `push`/`pop` functions does not work in O(1). But I would expect that.
Could you please evaluate the perfromance for `push`/`pop`? Leave the comment here.

Could you please improve the implementation?

Comment (TatsianaGouteeva) [reply Cocoaheads School_Lesson2.playground/Contents.swift]: But in our task only the function minimumElement should work in O(1).
Am I wrong?

File: Cocoaheads School_Lesson2.playground/Contents.swift (hunk 1 above)
Comment (TatsianaGouteeva) [reply Cocoaheads School_Lesson2.playground/Contents.swift]: done

File: Cocoaheads School_Lesson2.playground/Contents.swift (hunk 2 above)
Comment (dersim-davaod) [reply Cocoaheads School_Lesson2.playground/Contents.swift]: > But in our task only the function minimumElement should work in O(1).
Am I wrong? :)

You're right - you were explicitly asked to have O(1) for `minimumElement` only.

But what is the expected performance for stack's `push`/`pop` functions (see Readings sections for HW#2 to find the details)?
You have to meet these expectations + in addition, `minimumElement` should work in O(1).

Comment (TatsianaGouteeva) [reply Cocoaheads School_Lesson2.playground/Contents.swift]: All these operations of Stack should be O(1). That's why I rewrote functions, please, review my code.

File: Cocoaheads School_Lesson2.playground/Contents.swift [hunk 3]
```diff
@@ -0,0 +1,43 @@
+class Stack {
+    var array = [Int]()
+    var minElement: Int?
```
Comment (dersim-davaod) [inline Cocoaheads School_Lesson2.playground/Contents.swift]: I my code I can do:
```swift
let s = Stack()
s.minElement = 10
s.minElement
```

It returns `10` as a min element despite the stack is empty.

That does not seemed right. Please fix it.

File: Cocoaheads School_Lesson2.playground/Contents.swift [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    var minValues = [Int]()
```
Comment (dersim-davaod) [inline Cocoaheads School_Lesson2.playground/Contents.swift]: Could you please explain why you have to have an array?

File: Cocoaheads School_Lesson2.playground/Contents.swift (hunk 3 above)
Comment (TatsianaGouteeva) [reply Cocoaheads School_Lesson2.playground/Contents.swift]: done

File: Cocoaheads School_Lesson2.playground/Contents.swift (hunk 4 above)
Comment (TatsianaGouteeva) [reply Cocoaheads School_Lesson2.playground/Contents.swift]: By using an array we can show how Stack works.
Like an array Stack has common functions:
push, which adds an element to the collection, and
pop, which removes the most recently added element that was not yet removed.

Comment (dersim-davaod) [reply Cocoaheads School_Lesson2.playground/Contents.swift]: I meant: why you need `minValues` array to store the min value? You can use just a single variable.

File: Cocoaheads School_Lesson2.playground/Contents.swift [hunk 5]
```diff
@@ -1,43 +1,63 @@
...
-    }
-}
-
-class StackStatistics: Stack {
     func minimumElement() -> Int? {
-        return minElement
+        return minValues.last
+    }
+    
+    func theFirstMinimumElement() -> Int? {
```
Comment (dersim-davaod) [inline Cocoaheads School_Lesson2.playground/Contents.swift]: According to the task, your interface should not contain this method. Please remove it.

File: Cocoaheads School_Lesson2.playground/Contents.swift [hunk 6]
```diff
@@ -1,43 +1,63 @@
 class Stack {
-    var array = [Int]()
-    var minElement: Int?
-    var minValues = [Int]()
+    fileprivate var array = [Int]()
```
Comment (dersim-davaod) [inline Cocoaheads School_Lesson2.playground/Contents.swift]: For now this is more ok (this is not a mistake right now), but imagine you can have the following code in the current file:

```swift
let s = Stack()
s.array.append(10)
```

There's an issue with encapsulation. Could you please try to improve the implementation?

Comment (dersim-davaod) [reply Cocoaheads School_Lesson2.playground/Contents.swift]: As a result your interface allows you to access the private fields:

Screen Shot 2020-02-15 at 11 49 44 AM

File: Cocoaheads School_Lesson2.playground/Contents.swift (hunk 5 above)
Comment (TatsianaGouteeva) [reply Cocoaheads School_Lesson2.playground/Contents.swift]: done

File: Cocoaheads School_Lesson2.playground/Contents.swift (hunk 6 above)
Comment (TatsianaGouteeva) [reply Cocoaheads School_Lesson2.playground/Contents.swift]: Yes, I understood. In this case I should change the access level of my array. It should be private.

File: Cocoaheads School_Lesson2.playground/Contents.swift (hunk 4 above)
Comment (TatsianaGouteeva) [reply Cocoaheads School_Lesson2.playground/Contents.swift]: I used the array minValues  to store a minimum value for each  element in array Stack.
The array minValues I always update after pop/push functions, that's why I could get the minimum element in the stack at any time

Comment (dersim-davaod) [reply Cocoaheads School_Lesson2.playground/Contents.swift]: This ok for now :)

**Pull request merged**